cpp/test/mg/pca.cu

```diff
@@ -0,0 +1,214 @@
+#include <opg/linalg/gemm.hpp>
+#include <opg/matrix/matrix_utils.hpp>
+#include <matrix/matrix.cuh>
+#include "test_opg_utils.h"
+
+namespace MLCommon {
+namespace Test {
+namespace opg {
+
+struct PCAOpgParams {
+  int M;
+  int N;
+  int N_components;
+  ML::mg_solver algorithm;
+  std::vector<int> partSizes;
+  std::vector<int> ranksOwners;
+  Matrix::Layout layout;
+  unsigned long long int seed;
+};
+
+template <typename T>
+class PCAOpgTest : public testing::TestWithParam<PCAOpgParams> {
+ public:
+  void SetUp() {
+    params = GetParam();
+    handle = new ML::cumlHandle();
+    ML::initialize_mpi_comms(*handle, MPI_COMM_WORLD);
+
+    // Prepare resource
+    const ML::cumlHandle_impl& h = handle->getImpl();
+    const cumlCommunicator& comm = h.getCommunicator();
+    stream = h.getStream();
+    const std::shared_ptr<deviceAllocator> allocator = h.getDeviceAllocator();
+    cublasHandle_t cublasHandle = h.getCublasHandle();
+
+    myRank = comm.getRank();
+    totalRanks = comm.getSize();
+    Random::Rng r(params.seed + myRank);
+
+    CUBLAS_CHECK(cublasSetStream(cublasHandle, stream));
+
+    if (myRank == 0) {
+      std::cout << "Testing PCA of " << params.M << " x " << params.N
+                << " matrix" << std::endl;
+    }
+
+    // Prepare X matrix
+    std::vector<Matrix::RankSizePair*> totalPartsToRanks;
+    for (int i = 0; i < params.partSizes.size(); i++) {
+      Matrix::RankSizePair* rspt = new Matrix::RankSizePair(
+        params.ranksOwners[i] % totalRanks, params.partSizes[i]);
+      totalPartsToRanks.push_back(rspt);
+    }
+    Matrix::PartDescriptor desc(params.M, params.N, totalPartsToRanks,
+                                comm.getRank(), params.layout);
+    std::vector<Matrix::Data<T>*> inParts;
+    Matrix::opg::allocate(h, inParts, desc, myRank, stream);
+    Matrix::opg::randomize(h, r, inParts, desc, myRank, stream, T(10.0),
+                           T(20.0));
+    h.waitOnUserStream();
+
+    prmsPCA.n_rows = params.M;
+    prmsPCA.n_cols = params.N;
+    prmsPCA.n_components = params.N_components;
+    prmsPCA.whiten = false;
+    prmsPCA.n_iterations = 100;
+    prmsPCA.tol = 0.01;
+    prmsPCA.algorithm = params.algorithm;
+
+    components = (T*)allocator->allocate(
+      prmsPCA.n_components * prmsPCA.n_cols * sizeof(T), stream);
+
+    explained_var =
+      (T*)allocator->allocate(prmsPCA.n_components * sizeof(T), stream);
+
+    explained_var_ratio =
+      (T*)allocator->allocate(prmsPCA.n_components * sizeof(T), stream);
+
+    singular_vals =
+      (T*)allocator->allocate(prmsPCA.n_components * sizeof(T), stream);
+
+    mu = (T*)allocator->allocate(prmsPCA.n_cols * sizeof(T), stream);
+
+    noise_vars =
+      (T*)allocator->allocate(prmsPCA.n_components * sizeof(T), stream);
+
+    ML::PCA::opg::fit(*handle, inParts, desc, components, explained_var,
+                      explained_var_ratio, singular_vals, mu, noise_vars,
+                      prmsPCA, false);
```

Comment:
Reasonable test here could be to call inverse transform on the transformed output and verify that it matches the input. Can do it in this PR or create an issue. I'm okay with either.

Reply:
Calling inverse transform will depend on https://github.com/rapidsai/cuml/issues/2471, so I will create an issue to do that after this

Reply:
We should add a todo here for completeness and reference that issue.